### ISSUE TITLE
BUG,DISCUSSION:  What to do about reduction promotion/type resolution?

### ISSUE THREAD

**Issue by seberg** [numpy/numpy#18870]
This is mainly a few thoughts right now, but the promotion for reduction(-likes) is slightly weird.

Reductions might behave slightly different from `reduceat` (and `accumulate`, but accumulate is by definition more strict, as it requires all inputs and outputs to have the identical DType).  In practice this probably doesn't matter right now.

Also reduction on ufuncs that do not have a homogeneous signature are a bit strange, if they rely on meaning beyond `nonzero`:
```
In [14]: np.not_equal.reduce(np.ones(20))
Out[14]: False

In [15]: np.not_equal.reduce(np.ones(21))
Out[15]: True

In [16]: np.not_equal.reduce(np.ones(22))
Out[16]: False
```

I actually don't know the quite proper fix right now :(.  Adding explicit loops with `?d->?` signature is good, but we do have the funny fact that in this case, we don't have quite normal promotion, a fact that I never quite realized.  The issues is that one of the inputs is initially undefined.

Maybe that is simply fine, and we can define that an input dtype may be undefined in the case of reduction-like operations, and the promotion step has to deal with that possibility.

**Comment by rkern** [comment]
I don't see a problem. Not all operations that we expose necessarily make sense on all inputs. It's a dusty, cobwebby corner but not something that I think is particularly going to be a trap for anyone. I'm not sure there's any other result that would make sense, a result that people would think they were going to get which cause them to use this API with these inputs. This is exactly the result we get using the stdlib equivalents, for instance:

```
[~]                                                                                                                                                                                                       
|9> import functools, operator                                                                                                                                                                            

[~]                                                                                                                                                                                                       
|10> functools.reduce(operator.ne, [1.0]*20)                                                                                                                                                              
False

[~]                                                                                                                                                                                                       
|11> functools.reduce(operator.ne, [1.0]*21)                                                                                                                                                              
True

[~]                                                                                                                                                                                                       
|12> functools.reduce(operator.ne, [1.0]*22)                                                                                                                                                              
False
```

**Comment by seberg** [comment]
That is true, its not really a problem that there are some strange casts internally and the results don't make too much sense in those cases.

**Comment by seberg** [comment]
This is now bothering me for real, just because I reached the point where I need to figure it out.

I still agree that some weirdness can be expected, but note that using `np.ones` was unlucky. The example should be:
```
In [2]: import operator

In [3]: functools.reduce(operator.ne, [2.0]*20)
Out[3]: True

In [4]: functools.reduce(operator.ne, [2.0]*21)
Out[4]: True

In [5]: np.not_equal.reduce([2.0]*20)
Out[5]: False

In [6]: np.not_equal.reduce([2.0]*21)
Out[6]: True
```
or maybe more to the point:
```
np.equal.reduce([1., 2.]) == True
```
We cast before the operation, so things blow up. But `equal` would be weird anyway, since the first operation is `d,d->?` but after that its `?,d->?`, so there is no single loop that can handle it.

In any case this means that reduce currently has an extra "override the inputs" step which seems awkward in general, although it might be useful for logical functions.  I am seriously considering exploing the `?,d->?` idea a bit, even if we will need that fallback for a while in any case.

**Comment by seberg** [comment]
I am going to close this, as resolved, as it is resolved in the direction of an error in NumPy 1.22.  Until now, there was one bug report about it.  In that case, the user was better of having an error, because the code returned unexpected (and thus incorrect) results for them.

I am convinced this is the right choice, because there is a hidden unsafe cast to boolean involved (except for logical ufuncs, because there casting to boolean is by definition correct).

But if anyone has doubts, please note it here, in that case it may be good to look into having a 1.22.x release that allows it again.  (Personally, I may mainly consider allowing it again by adding a `casting="unsafe"` argument to `np.ufunc.reduce`.)
